tensorflow/c/tf_tensor.h

```diff
@@ -45,6 +45,24 @@ limitations under the License.
 extern "C" {
 #endif
 
+// Macro used to calculate struct size for maintaining ABI stability across 
+// different struct implementations. 
+#ifndef TF_OFFSET_OF_END
+#define TF_OFFSET_OF_END(TYPE, MEMBER) (offsetof(TYPE, MEMBER) + \
+    sizeof(((TYPE *)0)->MEMBER))
+#endif // TF_OFFSET_OF_END
+
+// Allocator Attributes used for tensor allocation. 
+typedef struct TF_AllocatorAttributes { 
```

Comment:
Out of curiosity, what is the design choice behind exposing the ABI of allocator attributes, rather than having an opaque pointer and setters (https://github.com/tensorflow/tensorflow/blob/7a4b89dbaf78503882dfeefa634c5bbff303d563/tensorflow/c/eager/c_api.h#L79-L85)? 

It looks like the original structure has an optimized bitmasking representation (https://github.com/tensorflow/tensorflow/blob/7a4b89dbaf78503882dfeefa634c5bbff303d563/tensorflow/core/framework/allocator.h#L386-L393); do we have any thoughts on re-using this instead of forking the C APi representation?

Reply:
There is a good way to check if a struct has a certain functionality or not (using `OFFSET`). 
Suppose instead we have a subset of functions and an opaque pointer. How can user check if a certain function is available in TF API? I guess we could keep incrementing a version. May be +@yisitu knows if there is a trade off compared to using struct size. One issue I can think of is that version needs to be updated manually but struct size is calculated with a macro.

Just passing a bitmask seems not extendable. What if we want to add another parameter which is not part of the bitmask?

Reply:
We discussed offline; the advantage of the struct size approach is forward compatibility (i.e. compiling against a newer libtensorflow.so's headers while linking against an older version), and the advantage of the opaque pointer approach is it allows us to keep a single representation for the AllocatorAttributes struct. 

Either approach SGTM :)

Reply:
Hi! Good question, I've listed the design considerations below. The thing to note is that these are trade-offs and not one approach is always better than the other.
* it is be cheaper to extend structure parameters than it would be to add another API or entry point. 
* legacy setter/getters have to remain behind and become maintenance and documentation overhead.
* given a dynamic library, to access the getters/setters (i.e. GetProcAddress/dlsym) you'll need a matching header file that exposes the exact function declarations.  With struct params you can use a non-matching header and support best effort functionality by inspection with offsetof.

Regarding bit masks, they are generally reserved to low-level close to hardware data structures. I suppose you can use enum flags as long as there are 62 values or less (reserve 0 value for unknown and max value for count).

Reply:
Thank you for the insights @yisitu ! After discussing with Brian, we decided to make it opaque pointer to match general structure of existing API.